### PR TITLE
[WIP] PreAuth endpoints.

### PR DESCRIPTION
**Pull request by YevhenShBolt**
Task - https://app.asana.com/0/1102891174882024/1109200201184567/f

### CODE REVIEW

File: Model/Api/CreateOrder.php [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+        $this->cartHelper = $cartHelper;
+    }
+
+    /**
+     * Pre-Auth hook: Create order.
+     *
+     * @api
+     *
+     * @param string $type - "order.create"
+     * @param array $order
```
Comment (mccottry) [inline Model/Api/CreateOrder.php]: The name here is a little misleading.  You have to variables that you are using: `$order` and `$createOrderData`.  Consider switching the names and then the names in context makes very clear self-documenting sense.

File: Helper/Order.php [hunk 2]
```diff
@@ -725,6 +726,158 @@ public function saveUpdateOrder($reference, $boltTraceId = null)
...
+
+        if (Hook::$fromBolt) {
+            $order->addStatusHistoryComment(
+                "BOLTPAY INFO :: THIS ORDER WAS CREATED VIA PRE-AUTH WEBHOOK"
+            );
+            // Send order confirmation email to customer.
+            // Emulate frontend area in order for email
+            // template to be loaded from the correct path
+            // even if run from the hook.
+            if (!$order->getEmailSent()) {
```
[the remark this answers is not in the document]
Comment (matt-thomason) [reply Helper/Order.php]: I agree - it is misleading to the customer if they fail auth after this email is sent.

File: Helper/Order.php [hunk 3]
```diff
@@ -725,6 +726,158 @@ public function saveUpdateOrder($reference, $boltTraceId = null)
...
+            $this->bugsnag->registerCallback(function ($report) use ($quote) {
+                $report->setMetaData([
+                    'Is Order Exist' => [
+                        'pre-auth order.create' => true,
+                        'parent quote ID' => $quote->getId(),
+                        'order reserved ID' => $quote->getReservedOrderId(),
+                    ]
+                ]);
+            });
+            throw new BoltException(
```
[the remark this answers is not in the document]
Comment (matt-thomason) [reply Helper/Order.php]: I like using `verify`. Even something like `verifyValidOrder` works.

File: Helper/Order.php [hunk 4]
```diff
@@ -725,6 +725,158 @@ public function saveUpdateOrder($reference, $boltTraceId = null)
...
+                __('Duplicate Order Creation Attempt. Order #: %1', $quote->getReservedOrderId()),
+                null,
+                \Bolt\Boltpay\Model\Api\CreateOrder::E_BOLT_ORDER_ALREADY_EXISTS
+            );
+        }
+
+        return false;
```
Comment (matt-thomason) [inline Helper/Order.php]: We don't need to return anything - it's never used.

File: Helper/Order.php [hunk 5]
```diff
@@ -725,6 +725,158 @@ public function saveUpdateOrder($reference, $boltTraceId = null)
...
+    }
+
+    /**
+     * @param $immutableQuote
+     * @param $transaction
```
Comment (matt-thomason) [inline Helper/Order.php]: Can you add types here as you did above?

File: Model/Api/CreateOrder.php [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+            /** @var \Magento\Sales\Model\Order $createOrderData */
+            $createOrderData = $this->orderHelper->preAuthCreateOrder($quote, $payload);
+
+            $this->sendResponse(200, [
+                'status'    => 'success',
+                'message'   => 'Order create was successful',
+                'display_id' => $createOrderData->getId(),
+                'total'      => $createOrderData->getGrandTotal() * 100,
+                'order_received_url' => '',
```
Comment (matt-thomason) [inline Model/Api/CreateOrder.php]: Is this url supposed to be blank?

File: Model/Api/CreateOrder.php (hunk 1 above)
Comment (YevhenShBolt) [reply Model/Api/CreateOrder.php]: These params come from the request and also implement Interface that should be equal to the request fields.

The request from Bolt: [{"type":"order.create","order":{},"currency":"USD"}]

File: Model/Api/CreateOrder.php (hunk 6 above)
Comment (YevhenShBolt) [reply Model/Api/CreateOrder.php]: Nope. I am working on it with Niraj now.

File: Helper/Order.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (YevhenShBolt) [reply Helper/Order.php]: I agree, removed.

File: Helper/Order.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (YevhenShBolt) [reply Helper/Order.php]: fixed.

File: Helper/Order.php (hunk 4 above)
Comment (YevhenShBolt) [reply Helper/Order.php]: fixed.

File: Model/Api/CreateOrder.php (hunk 6 above)
Comment (YevhenShBolt) [reply Model/Api/CreateOrder.php]: Fixed

File: Helper/Order.php (hunk 5 above)
Comment (YevhenShBolt) [reply Helper/Order.php]: What do you mean?

**Comment by YevhenShBolt** [comment]
Closing it. we have another branch for it https://github.com/BoltApp/bolt-magento2/pull/261

**Pull request closed without merging**